nexus/src/app/sagas/silo_create.rs

```diff
@@ -0,0 +1,196 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
```

Comment:
I originally had this as a transaction :) The problem I encountered is how to combine the transactional function's `conn`

```rust
pool.transaction(move |conn| {
 ...
})
 ```

with the fact that silo creation now requires calling a couple of Nexus / Datastore functions, without copying and pasting code into that transaction function.

For example, the saga calls `Nexus::silo_update_policy`, which calls `Datastore::role_assignment_replace_visible`. It's not clear to me how to expose what `role_assignment_replace_visible` does in a way that works when calling it from a transaction and on its own. One first thought is to move the bulk of what it does into a function that returns the appropriate `Query` type, then either calling execute with the pool or with the `conn` from the transaction. For this approach to work, anything that is called in this transaction has to be separated this way.

By putting it in a saga, there's no copying and splitting of functions. Of course, I could be missing a more straight-forward way to do this.

Reply:
I've been trying to make something like

```rust
pub struct DataStore {
    pool: Arc<dyn DbConnection + Send + Sync>,
}
```

build, with the hopes that something like this could work:

```rust
    pub async fn Nexus::silo_create(
        self: &Arc<Self>,
        opctx: &OpContext,
        new_silo_params: params::SiloCreate,
    ) -> CreateResult<db::model::Silo> {
        self.db_datastore
            .pool_authorized(opctx)
            .await?
            .transaction(move |conn| {
                let transaction_ds = DataStore {
                    pool: conn,
                };
                transaction_ds.silo_create(opctx, new_silo_params)
            })
    }
```

Something like this would cut down on the copy and pasting, and/or the splitting of functions that I mentioned in the previous comment. Problem is it doesn't build :)

Reply:
Yeah, agreed about that challenge and that's also how I'd go about trying to do it.  That is, my first thought is to try to have those functions user helper functions that return a `Query` or some type like it.  Maybe another option would be to have helper functions do the real work and have those functions accept the `conn` as an argument?  On the plus side I think the benefit here could be substantial.

Reply:
I was afraid I might be sending you on a wild goose chase so I wanted to try it out myself and came up with this:
https://github.com/jmpesp/omicron/compare/silo_groups...oxidecomputer:omicron:silo_groups_dap

I took the approach of moving the guts of the reused functions into helper functions that accept a `conn`.  This has its own ugliness.  On the plus side, it's a fair bit less code, and since it all goes out in one transaction, it's a lot less runtime work and there's no opportunity for Rust code to get any of the cleanup steps wrong.

It might be cleaner to change theses functions to just return the `Query` and execute that in the caller.  That would eliminate a lot of issues I had to work around here.  I didn't try that first only because I've had bad experiences writing functions that accept or return Diesel types.  But I think most of those problems involve trying to write generic functions -- these helpers might well be a lot more straightforward to have return `Query` instead.

Reply:
I did try to make this work by having these helper functions return an unexecuted query.  I spent over an hour with the diesel docs (including the guides), the example in pagination.rs, and rustc but could not figure out how to write down the return type of this function:

```rust
    fn silo_group_do_create(
        silo_group: SiloGroup,
    ) -> ??? {
        use db::schema::silo_group::dsl;

        diesel::insert_into(dsl::silo_group)
            .values(silo_group)
            .returning(SiloGroup::as_returning())
    }
```

such that I could use `get_result()` or `get_result_async()`.  Various approaches bottomed out with errors like:

```
error[E0277]: the trait bound `db::model::silo_group::SiloGroup: Queryable<(diesel::sql_types::Uuid, diesel::sql_types::Timestamptz, diesel::sql_types::Timestamptz, diesel::sql_types::Nullable<diesel::sql_types::Timestamptz>, diesel::sql_types::Uuid, diesel::sql_types::Text), Pg>` is not satisfied
    --> nexus/src/db/datastore/silo.rs:100:26
     |
100  |                         .get_result(conn)?;
     |                          ^^^^^^^^^^ the trait `Queryable<(diesel::sql_types::Uuid, diesel::sql_types::Timestamptz, diesel::sql_types::Timestamptz, diesel::sql_types::Nullable<diesel::sql_types::Timestamptz>, diesel::sql_types::Uuid, diesel::sql_types::Text), Pg>` is not implemented for `db::model::silo_group::SiloGroup`
     |
     = help: the trait `Queryable<(__ST0, __ST1, __ST2), __DB>` is implemented for `db::model::silo_group::SiloGroup`
     = note: required because of the requirements on the impl of `FromSqlRow<(diesel::sql_types::Uuid, diesel::sql_types::Timestamptz, diesel::sql_types::Timestamptz, diesel::sql_types::Nullable<diesel::sql_types::Timestamptz>, diesel::sql_types::Uuid, diesel::sql_types::Text), Pg>` for `db::model::silo_group::SiloGroup`
     = note: required because of the requirements on the impl of `CompatibleType<db::model::silo_group::SiloGroup, Pg>` for `(diesel::sql_types::Uuid, diesel::sql_types::Timestamptz, diesel::sql_types::Timestamptz, diesel::sql_types::Nullable<diesel::sql_types::Timestamptz>, diesel::sql_types::Uuid, diesel::sql_types::Text)`
     = note: required because of the requirements on the impl of `LoadQuery<'_, DTraceConnection<diesel::PgConnection>, db::model::silo_group::SiloGroup>` for `BoxedSelectStatement<'_, (diesel::sql_types::Uuid, diesel::sql_types::Timestamptz, diesel::sql_types::Timestamptz, diesel::sql_types::Nullable<diesel::sql_types::Timestamptz>, diesel::sql_types::Uuid, diesel::sql_types::Text), diesel::query_builder::FromClause<db::schema::silo_group::table>, Pg>`
```

I'm not sure if this is a result of our use of `diesel(embed)`?  I only say that because the number of elements in the expected tuple type (with `__ST0` and the like) matches the number of elements in the `SiloGroup` struct, but not the number of database columns (which is what's in the actual tuple type).  Anyway, [this is where I gave up trying](https://github.com/oxidecomputer/omicron/compare/silo_groups_dap...silo_groups_dap_queries).

Maybe @smklein has some ideas here?

Reply:
From a glance, it looks like the problem is coming from this function:

```rust
    pub(super) fn silo_group_do_create<'a, DB>(
        silo_group: SiloGroup,
    ) -> db::schema::silo_group::BoxedQuery<
        'a,
        DB,
        db::schema::silo_group::SqlType,
    > {
        use db::schema::silo_group::dsl;

        let x = diesel::insert_into(dsl::silo_group)
            .values(silo_group)
            .returning(SiloGroup::as_returning());
        x // XXX-dap
    }
```

rustc indicates that, although we've specified the return type to be a `BoxedQuery`, the type of `x` is an `InsertStatement`. In simpler terms, I think the concrete type we are returning is not boxed, but the type signature says that it is.

Taking a step back, it isn't clear to me if the return value in the type signature is actually what we want. Based on the callsites, it looks like you want this function to return a type where you can call either `.get_result` or `.get_result_async` - effectively, that https://docs.rs/diesel/2.0.0-rc.0/diesel/prelude/trait.RunQueryDsl.html#method.get_result is callable.

This seems possible by:
1. Actually boxing the result type
2. Ensuring it `impl`s `RunQueryDsl` and `LoadQuery`.

I'll attach the diff of "how-to-do-this" in a follow-up comment...

Reply:
I was able to make the `silo_groups_dap_queries` branch compile by using the following:

## Describe the return type based on "how we want to use the type"

As mentioned earlier, we want to be able to call `.get_result` (plus the async variant, but it has the same constraints).

I added a helper type specifying these bounds. If this is a pattern we use elsewhere, it would be easy to make this generic over the `SiloGroup` parameter, and make this a more generic `LoadableQuery` type.

```rust

// Add the following to `silo_group.rs`

pub(crate) trait SiloGroupCreateQuery :
    RunQueryDsl<crate::db::pool::DbConnection> +
    diesel::query_dsl::LoadQuery<'static, crate::db::pool::DbConnection, SiloGroup> {}

impl<T> SiloGroupCreateQuery for T
where
    T: RunQueryDsl<crate::db::pool::DbConnection> +
       diesel::query_dsl::LoadQuery<'static, crate::db::pool::DbConnection, SiloGroup>
{}
```

## Patch the return type of the helper function to return this boxed trait

Again, this just implies "the return value of this function should be a runnable query, which can load the `SiloGroup` type".

```rust
-    pub(super) fn silo_group_do_create<'a, DB>(
+    pub(super) fn silo_group_do_create(
         silo_group: SiloGroup,
-    ) -> db::schema::silo_group::BoxedQuery<
-        'a,
-        DB,
-        db::schema::silo_group::SqlType,
-    > {
+    ) -> Box<impl SiloGroupCreateQuery> {
```

## Actually box the returned type
```rust
-        x // XXX-dap
+        Box::new(x) // XXX-dap
```

Reply:
Thanks, @smklein!  That's very helpful.

How did you determine what the right return type should be?  I think the closest I got was when I tried returning `impl Query<SqlType = db::schema::silo_group::SqlType> + RunQueryDsl<DB>`.  The compiler told me I needed `QueryFragment`, then `QueryId`, then `LoadQuery`, but after `QueryId` I figured I must be on the wrong path (which I'd say I was).  I also tried returning the concrete type `InsertStatement` but found that wasn't possible (the concrete type is private to Diesel).  I also looked into a boxed result using `diesel::dsl::IntoBoxed`, but I got hung up again trying to figure out the trait bounds necessary for that.  That leads to my second question: do you know what `into_boxed()` is for, as opposed to `Box::new()`?

(I'm asking these things because I'm still trying to figure out how to more methodically approach the problem of writing functions that accept or return pieces of Diesel queries.)

Reply:
For the record, this works too (no helper type, no box):

```rust
    pub(super) fn silo_group_do_create(
        silo_group: SiloGroup,
    ) -> impl RunQueryDsl<DbConnection>
           + diesel::query_dsl::LoadQuery<'static, DbConnection, SiloGroup>
    {   
        use db::schema::silo_group::dsl;

        diesel::insert_into(dsl::silo_group)
            .values(silo_group)
            .returning(SiloGroup::as_returning())
    }   
```

Reply:
Ah, nice! That box-less solution seems great.

> How did you determine what the right return type should be? I think the closest I got was when I tried returning impl Query<SqlType = db::schema::silo_group::SqlType> + RunQueryDsl<DB>. The compiler told me I needed QueryFragment, then QueryId, then LoadQuery, but after QueryId I figured I must be on the wrong path (which I'd say I was). I also tried returning the concrete type InsertStatement but found that wasn't possible (the concrete type is private to Diesel).

When building queries, Diesel uses a *lot* of very specific types to encode all sorts of information - where the query is coming from, what type it's mapping into, if the return clause is set, etc, etc, etc. So I sorta expected it would be easier to "name the type based on how it is used" (traits) rather than "what it is" concretely.

This is what led me to https://docs.rs/diesel/2.0.0-rc.0/diesel/prelude/trait.RunQueryDsl.html#method.get_result - to call this function, we needed the `RunQueryDsl` + `LoadQuery` traits we ended up adding. I also saw that `InsertStatement` implements `RunQueryDsl`, and it likely also implemented `LoadQuery` (based on the diesel docs, which state "Its behavior is provided by `RunQueryDsl`) , so I attempted defining the output based on those two bounds.

> I also looked into a boxed result using diesel::dsl::IntoBoxed, but I got hung up again trying to figure out the trait bounds necessary for that. That leads to my second question: do you know what into_boxed() is for, as opposed to Box::new()?

FWIW, I didn't actually use `into_boxed()` here. I've *seen* it around Diesel, but based on https://docs.rs/diesel/2.0.0-rc.0/diesel/query_dsl/methods/trait.LoadQuery.html?search=into_boxed , I don't *think* it's actually a function that's callable on an InsertStatement. That's why I just used a regular `Box::new` - the boxing and return-by-trait pattern I used is independent of Diesel's boxing helpers.

Reply:
Looking at the documentation for, e.g., https://docs.rs/diesel/2.0.0-rc.0/diesel/query_builder/struct.UpdateStatement.html#method.into_boxed , it looks like the `into_boxed` function boxes the  `WHERE` clause.

Note that for a struct like `UpdateStatement`, it has a bunch of generic parameters. `into_boxed` appears to be boxing the `U` parameter (the `WHERE` clause specifically), not the whole `UpdateStatement` struct.

Reply:
silo creation is now a transaction as of 0aa76698